dynawo/sources/Models/CPP/Events/ModelAreaShedding/DYNModelAreaShedding.cpp

```diff
@@ -0,0 +1,332 @@
+  sizeZ_ = 1;  // delta started
+  sizeG_ = 1;  // activation of load shedding
+  sizeMode_ = 1;  // activation of load shedding
+
+  calculatedVars_.assign(nbCalculatedVars_, 0);
+}
+
+// evaluation of F(t,y,y') function
+
+void
+ModelAreaShedding::evalF(double /*t*/, propertyF_t type) {
+  if (type == DIFFERENTIAL_EQ)
+    return;
+  if (stateAreaShedding_ == NOT_STARTED) {  // shedding not started
+    for (int i = 0; i < nbLoads_; ++i) {
+      fLocal_[i * 2] = yLocal_[i * 2] - PRef_[i];
+      fLocal_[i * 2 + 1] = yLocal_[i * 2 + 1] - QRef_[i];
+    }
+  } else {  // shedding started
+    for (int i = 0; i < nbLoads_; ++i) {
+      fLocal_[i * 2] = yLocal_[i * 2] - PRef_[i] * (1 - deltaP_/100);
+      fLocal_[i * 2 + 1] = yLocal_[i * 2 + 1] - QRef_[i] * (1 - deltaQ_/100);
+    }
+  }
+}
+
+
+// evaluation of root functions
+
+void
+ModelAreaShedding::evalG(const double t) {
+  gLocal_[0] = (doubleEquals(t, deltaTime_) || (t - deltaTime_) > 0) ? ROOT_UP : ROOT_DOWN;
+}
+
+void
+ModelAreaShedding::setFequations() {
+  stringstream ss;
+  for (int i = 0; i < nbLoads_; ++i) {
+    ss.str("");
+    ss.clear();
+    ss << i;
+    fEquationIndex_[i * 2] = "PRef_" + ss.str();
+    fEquationIndex_[i * 2 + 1] = "QRef_" + ss.str();
+  }
+  assert(fEquationIndex_.size() == static_cast<size_t>(sizeF()) && "Model AreaShedding: fEquationIndex_.size() != fLocal_.size()");
+}
+
+void
+ModelAreaShedding::setGequations() {
+  gEquationIndex_[0] = "t >= deltaTime";
+
+  assert(gEquationIndex_.size() == static_cast<size_t>(sizeG()) && "Model AreaShedding: gEquationIndex.size() != gLocal_.size()");
+}
+
+void
+ModelAreaShedding::evalJt(const double /*t*/, const double /*cj*/, SparseMatrix& jt, const int rowOffset) {
+  static double dPOne = 1;
+  // whatever the state of the automaton, same Jacobian
+  for (int i = 0; i < nbLoads_; ++i) {  // 2 equations by loads
+    jt.changeCol();
+    jt.addTerm(i * 2 + rowOffset, dPOne);
+    jt.changeCol();
+    jt.addTerm(i * 2 + 1 + rowOffset, dPOne);
+  }
+}
+
+void
+ModelAreaShedding::evalJtPrim(const double /*t*/, const double /*cj*/, SparseMatrix& jt, const int /*rowOffset*/) {
+  // no differential equations
+  for (int i = 0; i < nbLoads_; ++i) {  // 2 equations by loads
+    jt.changeCol();
+    jt.changeCol();
+  }
+}
+
+// evaluation of discrete variables
+
+void
+ModelAreaShedding::evalZ(const double /*t*/) {
+  if (gLocal_[0] == ROOT_UP && stateAreaShedding_ != STARTED) {  // load shedding starts
+    DYNAddTimelineEvent(this, name(), LoadSheddingStarted);
+    zLocal_[0] = STARTED;
+    stateAreaShedding_ = STARTED;
+  }
+}
+
+
+void
+ModelAreaShedding::collectSilentZ(BitMask* silentZTable) {
+  silentZTable[0].setFlags(NotUsedInDiscreteEquations);
+}
+
+// evaluation of modes (alternatives) of F(t,y,y') functions
+
+modeChangeType_t
+ModelAreaShedding::evalMode(const double t) {
+  if ((started_ < 0 || doubleEquals(t, started_)) && t >= deltaTime_) {
+    started_ = t;
+    return ALGEBRAIC_MODE;
+  }
```

Comment:
At first glance, I don't see why we use started_ and have this additional attribute instead of using the stateAreaShedding for example.